File: tools/performance/perf_controller.cc
#include "drake/tools/performance/perf_controller.h"

#include <fcntl.h>
#include <unistd.h>

#include <cstdlib>
#include <cstring>
#include <stdexcept>

#include "drake/common/never_destroyed.h"

namespace drake {
namespace tools {
namespace performance {

namespace {

constexpr const char enable_cmd[] = "enable";
constexpr int enable_len = sizeof(enable_cmd);
constexpr const char disable_cmd[] = "disable";
constexpr int disable_len = sizeof(disable_cmd);
constexpr const char ack_cmd[] = "ack\n";
constexpr int ack_len = sizeof(ack_cmd);

int FdFromEnv(const char* env_key) {
  const char* name = std::getenv(env_key);
  if (name == nullptr) {
    return -1;
  }
  return open(name, O_RDWR);
}

}  // namespace

PerfController::PerfController() {
  ctl_fd_ = FdFromEnv("DRAKE_PERF_CTL_FIFO");
  ack_fd_ = FdFromEnv("DRAKE_PERF_ACK_FIFO");
  is_control_available_ = (ctl_fd_ >= 0 && ack_fd_ >= 0);
}

void PerfController::pause() {
  if (!is_control_available()) {
    return;
  }
  send_command(disable_cmd, disable_len);
}

void PerfController::resume() {
  if (!is_control_available()) {
    return;
  }
  send_command(enable_cmd, enable_len);
}

bool PerfController::is_control_available() const {
  return is_control_available_;
}

void PerfController::send_command(const char* command, int command_len) {
  if (!is_control_available()) {
    return;
  }
  int wrote_bytes = write(ctl_fd_, command, command_len);
  if (wrote_bytes < command_len) {
    throw std::runtime_error("`perf` command not completely written.");
  }
  char ack[5];
  int read_bytes = read(ack_fd_, ack, ack_len);
  if (read_bytes < ack_len) {
    throw std::runtime_error(
        "`perf` command acknowledgment not completely read.");
  }
  if (std::strcmp(ack, ack_cmd) != 0) {
    throw std::runtime_error("`perf` command acknowledgment not received.");
  }
}

PerfController& ThePerfController() {
  static never_destroyed<PerfController> controller;
  return controller.access();
}

}  // namespace performance
}  // namespace tools
}  // namespace drake

File: tools/performance/perf_controller.py
"""Instrumentation helpers for in-process control of profile sampling by Linux
`perf`. See README.md. for an overview.
"""

from contextlib import contextmanager
import os

_enable_cmd = "enable".encode('utf-8')
_disable_cmd = "disable".encode('utf-8')
_ack_cmd = "ack\n\0".encode('utf-8')


def _fd_from_env(env_key: str) -> int:
    name = os.environ.get(env_key)
    if not name:
        return -1
    return os.open(name, os.O_RDWR)


class _PerfController:
    """In-process controller class.

    This class is designed to work together with the top-level launch script
    perf_controlled_record.py. It looks for the names of two FIFO files via the
    environment variables DRAKE_PERF_CTL_FIFO and DRAKE_PERF_ACK_FIFO. If those
    are missing, or the files can't be opened, then control will be
    unavailable.

    No attempt is made to make this class thread-safe, so be careful to avoid
    issuing commands from multiple threads.
    """

    def __init__(self):
        self._ctl_fd = _fd_from_env("DRAKE_PERF_CTL_FIFO")
        self._ack_fd = _fd_from_env("DRAKE_PERF_ACK_FIFO")
        self._is_control_available = (self._ctl_fd >= 0 and self._ack_fd >= 0)

    def pause(self):
        """Turn sampling off."""
        if not self.is_control_available():
            return
        self._send_command(_disable_cmd)

    def resume(self):
        """ Turn sampling on."""
        if not self.is_control_available():
            return
        self._send_command(_enable_cmd)

    def is_control_available(self):
        """Returns True iff sampling control is available."""
        return self._is_control_available

    def _send_command(self, command):
        if not self.is_control_available():
            return
        wrote_bytes = os.write(self._ctl_fd, command)
        if wrote_bytes != len(command):
            raise RuntimeError("`perf` command not completely written.")
        got = os.read(self._ack_fd, len(_ack_cmd))
        if got != _ack_cmd:
            raise RuntimeError(
                f"`perf` command acknowledgment not received:"
                f" {got} != {_ack_cmd}")


_the_perf_controller = _PerfController()


def ThePerfController():
    """Returns the singleton PerfController."""
    return _the_perf_controller


@contextmanager
def scoped_perf_sampling():
    """Context manager to turn sampling on for the duration
    of a section of code.
    """
    ThePerfController().resume()
    try:
        yield
    finally:
        ThePerfController().pause()
